test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,512 @@
+		Expect(framework.ReportSpecStart(cluster)).To(Succeed())
+	})
+
+	E2EAfterAll(func() {
+		Expect(cluster.TriggerDeleteNamespace(TestNamespace)).To(Succeed())
+		Expect(cluster.DeleteKuma()).To(Succeed())
+	})
+
+	It("should deploy the mesh", func() {
+		// just to see stabilized stats before we go further
+		Expect(true).To(BeTrue())
+	})
+
+	It("should deploy mesh wide policy", func() {
+		policy := `
+apiVersion: kuma.io/v1alpha1
+kind: MeshRateLimit
+metadata:
+  name: mesh-rate-limit
+  namespace: kong-mesh-system
+spec:
+  targetRef:
+    kind: Mesh
+    proxyTypes:
+      - Sidecar
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        local:
+          http:
+            requestRate:
+              num: 10000
+              interval: 1s
+            onRateLimit:
+              status: 429
+`
+		Expect(cluster.Install(YamlK8s(policy))).To(Succeed())
+	})
+
+	Context("load", func() {
+		var cpu int
+		var memory int
+		var numServices = 5
+		var instancesPerService = 1
+		var goDebugAdded = false
+		var goMemLimitAdded = false
+
+		adjustResource := func(miliCPU, memMega int, addGoMemLimitEnv bool) {
+			printLog("adjusting control plane resource limits to cpu %dm, memory %dMi\n", miliCPU, memMega)
+
+			patchJson := `[
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/requests/cpu", "value": "%dm"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/requests/memory", "value": "%dMi"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/cpu", "value": "%dm"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/memory", "value": "%dMi"}`
+
+			if !goDebugAdded {
+				goDebugAdded = true
+				patchJson += `,{"op": "add", "path": "/spec/template/spec/containers/0/env/-", "value": {"name": "GODEBUG", "value":"gctrace=1"}}`
+			}
+
+			if addGoMemLimitEnv {
+				memLimit := memMega - 50
+				if memLimit < 10 {
+					memLimit = 10
+				}
+
+				if !goMemLimitAdded {
+					patchJson += fmt.Sprintf(`,{"op": "add", "path": "/spec/template/spec/containers/0/env/-", "value": {"name": "GOMEMLIMIT", "value":"%dMiB"}}`, memLimit)
+					goMemLimitAdded = true
+				} else {
+					//  get the existing env array, and update the existing GOMEMLIMIT
+					out, err := k8s.RunKubectlAndGetOutputE(
+						cluster.GetTesting(),
+						cluster.GetKubectlOptions(Config.KumaNamespace),
+						"get", "deployment", Config.KumaServiceName,
+						"-o=jsonpath='{.spec.template.spec.containers[0].env}'")
+					Expect(err).ToNot(HaveOccurred())
+					var jsonEnvArray []map[string]interface{}
+					err = yaml.Unmarshal([]byte(out), &jsonEnvArray)
+					Expect(err).ToNot(HaveOccurred())
+					var idx = -1
+					for i, m := range jsonEnvArray {
+						if m["name"] == "GOMEMLIMIT" {
+							idx = i
+							break
+						}
+					}
+
+					op := "replace"
+					idxStr := fmt.Sprintf("%d", idx)
+					if idx == -1 {
+						op = "add"
+						idxStr = "-"
+					}
```

Comment:
Do we need to handle -1 idx and `-` even? Can we simplify the code by assuming that it was added already because `goMemLimitAdded` tracks this

Reply:
`goMemLimitAdded` means whether we want to add the env var in this call, while this `idx` tracks if the env var exists in the deployment.